src/content/docs/en/guides/environment-variables.mdx

```diff
@@ -9,38 +9,240 @@ Astro supports 2 main ways for handling environment variables: the `astro:env` m
+const data = await fetch(`${API_URL}/users`, {
+  method: "GET",
+  headers: {
+    "Content-Type": "application/json",
+    "Authorization": `Bearer ${API_SECRET_TOKEN}`
+  },
+});
+---
+
+<script>
+import { API_URL } from "astro:env/client"
+fetch(`${API_URL}/ping`)
+</script>
+```
 
-TODO: where to talk about integrations & adapters?
+**TODO: where to talk about integrations & adapters?**
 
 ### Define your schema
 
-- envField helper
-- field types + validations
+To configure a schema, add the `env.schema` option to your Astro config:
+
+```js title="astro.config.mjs" ins={4-8}
+import { defineConfig } from 'astro/config'
+
+export default defineConfig({
+  env: {
+    schema: {
+      // ...
+    }
+  }
+})
+```
+
+You can then register variables using the `envField` helper:
+
+```js title="astro.config.mjs" ins={3,9-11}
+import {
+   defineConfig,
+   envField
+} from 'astro/config'
+
+export default defineConfig({
+  env: {
+    schema: {
+      API_URL: envField.string({ context: "client", access: "public", optional: true }),
+      PORT: envField.number({ context: "server", access: "public", default: 4321 }),
+      API_SECRET: envField.string({ context: "server", access: "secret" }),
+    }
+  }
+})
+```
 
-### Typegen
+Types will be generated for you when running `astro dev` or `astro build`, but you can run `astro sync` to generate types only.
 
 ### Variables types
 
-- client/public (accessible on client+server, only use not sensitive data)
-- client/secret (impossible)
-- server/public (accessible on server, can remain in build output)
-- server/secret (accessible on server, no trace anywhere, validated at runtime unless validateSecrets is used)
+There are three kinds of environment variables, determined by the combination of `context` (client or server) and `access` (secret or public) settings defined in your schema:
+
+- **Public client variables**: These variables end up in both your final client and server bundles, and can be accessed from both client and server through the `astro:env/client` module:
+
+   ```js
+   import { API_URL } from "astro:env/client"
+   ```
+
+- **Public server variables**: These variables end up in your final server bundle and can be accessed on the server through the `astro:env/server` module:
+
+   ```js
+   import { PORT } from "astro:env/server"
+   ```
+
+- **Secret server variables**: These variables are not part of your final bundle and can be accessed on the server through the `astro:env/server` module:
+
+   ```js
+   import { API_SECRET } from "astro:env/server"
+   ```
+
+   Worth nothing that by default, secrets are only validated at runtime. To validate secrets on start, check [validateSecrets](#validatesecrets-option).
+
+**Note:** Secret client variables are not supported because there is no safe way to send this data to the client. Therefore, it is not possible to configure both `context: "client"` and `access: "secret"` in your schema.
+
+### Data types
```

Comment:
A lot of this is starting to get quite API referencey. Not everything is appropriate for a guide page. We should have an `astro:env` section on the API reference page for this content.